proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,654 @@
+        name: servercert
+        namespace: # namespace of the configmap, if omitted assumes namespace of this resource
+      trustOptions:
+       clientAuth: REQUIRED
+    infrastructure:
+      # Controls that let the Developer influence the annotations/labels created on the ingress objects (Service, OpenShift Routes, TLSRoutes etc).
+      # If an annotation or label supplied here duplicates one provided by the ProxyIngress, the one defined by the ProxyIngress takes precendence. 
+      # The specific use case for this is https://docs.openshift.com/container-platform/4.17/security/certificates/service-serving-certificate.html#add-service-certificate_service-serving-certificate
+      # where the developer wishes their VC to use Certificates created by OpenShift (useful for ClusterIP type TLS).
+      annotations:
+         x: y
+      labels:
+         x: y
+
+  # Points to the cluster being proxied.  Can either be Strimzi Kafka resource or endpoint details.
+  targetCluster:
+    # one of: resourceRef or resource
+    resourceRef:
+      kind: Kafka  # must be Kafka
+      group: strimzi.io # must be strimzi.io
+      name: my-cluster
+      listenerName: listener # name of strimzi listener 
+    resource:
+       bootstrap: bootstrap:9092
+       protocol: TCP|TLS
+       nodeIdRanges:
+      - name: mybrokers
+        range:
+          startInclusive: 0
+          endExclusive: 3
+    tls:
+      # Optional - client auth
+      certificateRef:
+        kind: Secret # if present must be Secret, otherwise defaulted to Secret
+        group: ""  # if present must be "", otherwise defaulted to ""
+        name: servercert
+        namespace: # namespace of the secret, if omitted assumes namespace of this resource
+      # Optional - peer trust
+      trustAnchorRefs:
+       - kind: ConfigMap # if present must be ConfigMap, otherwise defaulted to ConfigMap
+         group: ""  # if present must be "", otherwise defaulted to ""
+         name: servercert
+         namespace: # namespace of the configmap, if omitted assumes namespace of this resource
+
+    # ordered list of filters to be used by the virtualcluster
+    filterRefs:
+    - group: filter.kroxylicious.io
+      kind: Filter
+      name: encryption
+status:
+   # overall status
+   conditions:
+   - status: x
+     type: y
+     reason: r 
+   # per ingress status 
+   ingresses:
+   - name: myclusterip
+     bootstrap: xxx.local:9082   # the bootstrap address
+     protocol: TCP|TLS
+     conditions:
+     - ...
+   # per filter status 
+   filters:
+   - name: myclusterip
+     conditions:
+     - ...
+```
+
+### Filter
+
+The VirtualCluster CR is the responsibilty of the the Developer - As per current implementation.
+
+```yaml
+apiVersion: filter.kroxylicious.io/v1alpha1
+kind: Filter
+metadata:
+  name: myfilter
+spec:
+  type: io.kroxylicious.filter.encryption.RecordEncryption
+  config:
+    kms: Foo
+    kmsConfig: {}
+    selector: Bar
+    selectorConfig: {}
```

Comment:
What is your thinking here around the schema for `Kind: filter`? Are all filters going to have standard  generic schema? e.g. `type` + arbitrary config? 

Or should we be offering a higher level CR for Filters shipped by the project? 

I started wondering about this from the angle of what happens if we want to have multiple filter instances pointed at the same KMS possibly even a corporate standard KMS instance that the infrastructure persona manages.  Should the filter also support a `kmsRef` or just a `configRef`? that points to things configured once?

Reply:
I haven't thought about this aspect at all this proposal!    The Filter definition you see is what we have on main right now.

In https://github.com/kroxylicious/design/blob/playing-with-operator-ingress-ideas/proposals/kroxylicious-operator-api-alternative.md#phased-high-level-implementation-plan I've said 'allow filters to accept secrets'.  Personally I feel that @robobario ["secret notation"](https://github.com/kroxylicious/kroxylicious/discussions/1613#discussioncomment-11328768) is good enough.    I don't think we want to start thinking about making KMSs toplevel right now.